### PR TITLE
Timer endpoints

### PR DESCRIPTION
**Pull request by lydiagu**

### CODE REVIEW

File: orchestra/api_urls.py [hunk 1]
```diff
@@ -45,6 +48,12 @@
     url(r'^interface/time_entries/$',
         time_entries, name='time_entries'),
 
+    url(r'^interface/timer/start/$', start_timer, name='start_timer'),
+
```
Comment (marcua) [inline orchestra/api_urls.py]: you can ditch the newlines between the endpoints

File: orchestra/views.py [hunk 2]
```diff
@@ -233,6 +237,58 @@ def time_entries(request):
...
+            timer = time_tracking.start_timer(worker,
+                                              assignment_id=assignment_id)
+            serializer = TaskTimerSerializer(timer)
+            return serializer.data
+    except TaskAssignment.DoesNotExist:
+        raise BadRequest('Worker is not assigned to this task id.')
+    except TimerError as e:
+        raise BadRequest(e)
+    except Exception as e:
+        logger.error(e, exc_info=True)
```
Comment (marcua) [inline orchestra/views.py]: logger.exception, no need for exc_info

File: orchestra/views.py [hunk 3]
```diff
@@ -233,6 +237,58 @@ def time_entries(request):
...
+    worker = Worker.objects.get(user=request.user)
+    try:
+        if request.method == 'POST':
+            time_entry = time_tracking.stop_timer(worker)
+            serializer = TimeEntrySerializer(time_entry)
+            return serializer.data
+    except TimerError as e:
+        raise BadRequest(e)
+    except Exception as e:
+        logger.error(e, exc_info=True)
```
Comment (marcua) [inline orchestra/views.py]: logger.exception, no need for exc_info

File: orchestra/views.py [hunk 4]
```diff
@@ -233,6 +237,58 @@ def time_entries(request):
...
+@json_view
+@login_required
+def get_timer(request):
+    worker = Worker.objects.get(user=request.user)
+    try:
+        if request.method == 'GET':
+            duration = time_tracking.get_timer_current_duration(worker)
+            return str(duration)
+    except Exception as e:
+        logger.error(e, exc_info=True)
```
Comment (marcua) [inline orchestra/views.py:288]: logger.exception, no need for exc_info

File: orchestra/tests/test_views.py [hunk 5]
```diff
@@ -131,3 +150,81 @@ def test_time_entries_post_data_invalid(self, mock_func, mock_logger):
...
+        resp = self.request_client.post(
+            reverse('orchestra:orchestra:start_timer'),
+            data=json.dumps({'assignment': '11'}),
+            content_type='application/json')
+        self._verify_bad_request(resp,
+                                 'Worker is not assigned to this task id.')
+
+    @patch('orchestra.views.time_tracking.start_timer',
+           side_effect=TimerError('test'))
+    def test_start_timer_timer_already_running(self, mock_start):
```
Comment (marcua) [inline orchestra/tests/test_views.py]: if this is testing for already running, why do you have to mock an exception on start_timer?

File: orchestra/tests/test_views.py [hunk 6]
```diff
@@ -131,3 +150,81 @@ def test_time_entries_post_data_invalid(self, mock_func, mock_logger):
...
+        user = UserFactory(username='test_user')
+        self.worker = WorkerFactory(user=user)
+        self.request_client = RequestClient()
+        self.request_client.login(username=self.worker.user.username,
+                                  password='defaultpassword')
+        self.time = timezone.now()
+
+    @patch('orchestra.views.time_tracking.start_timer')
+    def test_start_timer(self, mock_start):
+        timer = TaskTimer(worker=self.worker, start_time=self.time)
```
Comment (marcua) [inline orchestra/tests/test_views.py:169]: if there is a start_time already set, shouldn't we get an exception saying the timer is already running?

Comment (lydiagu) [inline orchestra/tests/test_views.py:169]: I'm mocking start_timer. timer is the return value of the mock

File: orchestra/tests/test_views.py (hunk 5 above)
Comment (lydiagu) [inline orchestra/tests/test_views.py]: i'm testing how the endpoint handles when the mocked function throws a `TimerError` exception. TimerEndpointTests is only testing the logic in views.py, not the logic in utils.time_tracking. That code is tested in test_time_tracking

File: orchestra/views.py (hunk 2 above)
Comment (lydiagu) [inline orchestra/views.py]: done

File: orchestra/views.py (hunk 3 above)
Comment (lydiagu) [inline orchestra/views.py]: done

File: orchestra/views.py (hunk 4 above)
Comment (lydiagu) [inline orchestra/views.py:288]: done

**Pull request merged**